File: src-electron/apk-parser.ts
import { readFileSync } from "node:fs";
import { inflateRawSync } from "node:zlib";

/**
 * Extract the package name from an APK by parsing the binary AndroidManifest.xml.
 * Uses only Node.js built-in modules (no external ZIP library).
 */
export function extractPackageName(apkPath: string): string | null {
  try {
    const manifest = readZipEntry(apkPath, "AndroidManifest.xml");
    if (!manifest) return null;
    return parsePackageName(manifest);
  } catch {
    return null;
  }
}

// ---------------------------------------------------------------------------
// Minimal ZIP reader — extracts a single entry by name
// ---------------------------------------------------------------------------

function readZipEntry(filePath: string, entryName: string): Buffer | null {
  const buf = readFileSync(filePath);

  // Find End of Central Directory record (signature 0x06054b50)
  let eocd = -1;
  for (let i = buf.length - 22; i >= 0; i--) {
    if (buf.readUInt32LE(i) === 0x06054b50) {
      eocd = i;
      break;
    }
  }
  if (eocd === -1) return null;

  const cdOffset = buf.readUInt32LE(eocd + 16);
  const cdCount = buf.readUInt16LE(eocd + 10);

  // Walk Central Directory entries
  let pos = cdOffset;
  for (let i = 0; i < cdCount; i++) {
    if (buf.readUInt32LE(pos) !== 0x02014b50) break;

    const compSize = buf.readUInt32LE(pos + 20);
    const nameLen = buf.readUInt16LE(pos + 28);
    const extraLen = buf.readUInt16LE(pos + 30);
    const commentLen = buf.readUInt16LE(pos + 32);
    const localOffset = buf.readUInt32LE(pos + 42);
    const name = buf.subarray(pos + 46, pos + 46 + nameLen).toString("utf8");

    if (name === entryName) {
      // Read Local File Header
      const lh = localOffset;
      if (buf.readUInt32LE(lh) !== 0x04034b50) return null;
      const method = buf.readUInt16LE(lh + 8);
      const lhNameLen = buf.readUInt16LE(lh + 26);
      const lhExtraLen = buf.readUInt16LE(lh + 28);
      const dataStart = lh + 30 + lhNameLen + lhExtraLen;
      const raw = buf.subarray(dataStart, dataStart + compSize);

      if (method === 0) return raw as Buffer;
      if (method === 8) return inflateRawSync(raw);
      return null;
    }

    pos += 46 + nameLen + extraLen + commentLen;
  }
  return null;
}

// ---------------------------------------------------------------------------
// Minimal Android binary XML parser — only extracts `<manifest package="...">`
// ---------------------------------------------------------------------------

function readU16(b: Buffer, o: number): number {
  return b.readUInt16LE(o);
}

function readU32(b: Buffer, o: number): number {
  return b.readUInt32LE(o);
}

function parseStringPool(buf: Buffer, cs: number): string[] {
  const count = readU32(buf, cs + 8);
  const flags = readU32(buf, cs + 16);
  const stringsStart = readU32(buf, cs + 20);
  const isUtf8 = (flags & 0x100) !== 0;
  const offsetsBase = cs + 28;
  const dataBase = cs + stringsStart;

  const strings: string[] = [];
  for (let i = 0; i < count; i++) {
    const off = dataBase + readU32(buf, offsetsBase + i * 4);
    if (off >= buf.length) {
      strings.push("");
      continue;
    }

    if (isUtf8) {
      let p = off;
      // char count (1-2 bytes)
      if (buf[p] & 0x80) p += 2;
      else p += 1;
      // byte count (1-2 bytes)
      let bc: number;
      if (buf[p] & 0x80) {
        bc = ((buf[p] & 0x7f) << 8) | buf[p + 1];
        p += 2;
      } else {
        bc = buf[p];
        p += 1;
      }
      strings.push(buf.subarray(p, Math.min(p + bc, buf.length)).toString("utf8"));
    } else {
      const cc = readU16(buf, off);
      const start = off + 2;
      const codes: number[] = [];
      for (let j = 0; j < cc; j++) {
        const idx = start + j * 2;
        if (idx + 1 < buf.length) codes.push(readU16(buf, idx));
      }
      strings.push(String.fromCharCode(...codes));
    }
  }
  return strings;
}

function parsePackageName(buf: Buffer): string | null {
  if (buf.length < 8 || readU32(buf, 0) !== 0x00080003) return null;
  if (readU16(buf, 8) !== 0x0001) return null;

  const spSize = readU32(buf, 12);
  const strings = parseStringPool(buf, 8);

  let pos = 8 + spSize;
  while (pos + 8 <= buf.length) {
    const type = readU16(buf, pos);
    const size = readU32(buf, pos + 4);
    if (size === 0) break;

    // START_ELEMENT = 0x0102
    if (type === 0x0102 && pos + 36 <= buf.length) {
      const nameIdx = readU32(buf, pos + 20);
      if (nameIdx < strings.length && strings[nameIdx] === "manifest") {
        const attrCount = readU16(buf, pos + 28);
        for (let a = 0; a < attrCount; a++) {
          const ao = pos + 36 + a * 20;
          if (ao + 20 > buf.length) break;
          const anIdx = readU32(buf, ao + 4);
          if (anIdx < strings.length && strings[anIdx] === "package") {
            const rawVal = readU32(buf, ao + 8);
            if (rawVal < strings.length) return strings[rawVal];
            if (buf[ao + 15] === 3) {
              const td = readU32(buf, ao + 16);
              if (td < strings.length) return strings[td];
            }
          }
        }
        return null;
      }
    }
    pos += size;
  }
  return null;
}

File: src-tauri/src/apk_parser.rs
use std::io::Read;

/// Extract the package name from an APK file by parsing AndroidManifest.xml binary XML.
pub fn extract_package_name(apk_path: &str) -> Result<String, String> {
    let file =
        std::fs::File::open(apk_path).map_err(|e| format!("无法打开 APK: {}", e))?;
    let mut archive =
        zip::ZipArchive::new(file).map_err(|e| format!("无效的 APK 文件: {}", e))?;
    let mut manifest = archive
        .by_name("AndroidManifest.xml")
        .map_err(|_| "APK 中未找到 AndroidManifest.xml".to_string())?;
    let mut buf = Vec::new();
    manifest
        .read_to_end(&mut buf)
        .map_err(|e| format!("读取 Manifest 失败: {}", e))?;
    parse_package_name(&buf)
}

// ---------------------------------------------------------------------------
// Minimal Android binary XML parser — only extracts the `package` attribute
// from the `<manifest>` element.
// ---------------------------------------------------------------------------

fn read_u16(d: &[u8], o: usize) -> u16 {
    u16::from_le_bytes([d[o], d[o + 1]])
}

fn read_u32(d: &[u8], o: usize) -> u32 {
    u32::from_le_bytes([d[o], d[o + 1], d[o + 2], d[o + 3]])
}

/// Parse the string pool chunk and return all strings.
fn parse_string_pool(data: &[u8], cs: usize) -> Result<Vec<String>, String> {
    let string_count = read_u32(data, cs + 8) as usize;
    let flags = read_u32(data, cs + 16);
    let strings_start = read_u32(data, cs + 20) as usize;
    let is_utf8 = (flags & 0x100) != 0;

    let offsets_start = cs + 28;
    let abs_strings_start = cs + strings_start;

    let mut strings = Vec::with_capacity(string_count);
    for i in 0..string_count {
        let offset = abs_strings_start + read_u32(data, offsets_start + i * 4) as usize;
        if offset >= data.len() {
            strings.push(String::new());
            continue;
        }

        if is_utf8 {
            let mut pos = offset;
            // char count (1-2 bytes)
            if data[pos] & 0x80 != 0 {
                pos += 2;
            } else {
                pos += 1;
            }
            // byte count (1-2 bytes)
            let byte_count = if data[pos] & 0x80 != 0 {
                let hi = (data[pos] & 0x7F) as usize;
                let lo = data[pos + 1] as usize;
                pos += 2;
                (hi << 8) | lo
            } else {
                let c = data[pos] as usize;
                pos += 1;
                c
            };
            let end = (pos + byte_count).min(data.len());
            strings.push(String::from_utf8_lossy(&data[pos..end]).to_string());
        } else {
            let char_count = read_u16(data, offset) as usize;
            let start = offset + 2;
            let mut u16s = Vec::with_capacity(char_count);
            for j in 0..char_count {
                let idx = start + j * 2;
                if idx + 1 < data.len() {
                    u16s.push(read_u16(data, idx));
                }
            }
            strings.push(String::from_utf16_lossy(&u16s));
        }
    }
    Ok(strings)
}

/// Walk the binary XML to find `<manifest package="...">` and return the value.
fn parse_package_name(data: &[u8]) -> Result<String, String> {
    if data.len() < 8 || read_u32(data, 0) != 0x0008_0003 {
        return Err("非二进制 XML 格式".into());
    }

    // String pool is the first chunk (right after the 8-byte file header)
    if read_u16(data, 8) != 0x0001 {
        return Err("未找到字符串池".into());
    }
    let sp_chunk_size = read_u32(data, 12) as usize;
    let strings = parse_string_pool(data, 8)?;

    // Scan remaining chunks for the first START_ELEMENT (type 0x0102)
    let mut pos = 8 + sp_chunk_size;
    while pos + 8 <= data.len() {
        let chunk_type = read_u16(data, pos);
        let chunk_size = read_u32(data, pos + 4) as usize;
        if chunk_size == 0 {
            break;
        }

        if chunk_type == 0x0102 && pos + 36 <= data.len() {
            let name_idx = read_u32(data, pos + 20) as usize;
            if name_idx < strings.len() && strings[name_idx] == "manifest" {
                let attr_count = read_u16(data, pos + 28) as usize;
                for a in 0..attr_count {
                    let ao = pos + 36 + a * 20;
                    if ao + 20 > data.len() {
                        break;
                    }
                    let attr_name_idx = read_u32(data, ao + 4) as usize;
                    if attr_name_idx < strings.len() && strings[attr_name_idx] == "package" {
                        // Try raw string value
                        let raw_val = read_u32(data, ao + 8) as usize;
                        if raw_val < strings.len() {
                            return Ok(strings[raw_val].clone());
                        }
                        // Fallback: typed value (dataType 3 = string)
                        if data[ao + 15] == 3 {
                            let typed_data = read_u32(data, ao + 16) as usize;
                            if typed_data < strings.len() {
                                return Ok(strings[typed_data].clone());
                            }
                        }
                    }
                }
                return Err("manifest 元素未找到 package 属性".into());
            }
        }
        pos += chunk_size;
    }
    Err("未找到 manifest 元素".into())
}
